### ISSUE TITLE
stability: ungraceful shutdown of last node

### ISSUE THREAD

**Issue by mberhault** [cockroachdb/cockroach#5279]
Build sha: 2132e353ded989829d7c92a09e6e1745b02678c7

Four nodes with little traffic (photos app was stopped about a minute before shutdown attempt).
First three nodes shutdown fairly quickly (seconds).
The last node eventually performed a hard shutdown after reaching the time limit.

Nodes were sent `SIGTERM` in numerical order (node0, node1, node2, node3)

Log (relevant excerpts and full):
Node 0:
```
I160315 20:50:37.292840 server/status/runtime.go:140  runtime stats: 100 goroutines, 15.85MiB active, 1272.60cgo/sec, 0.03/0.00 %(u/s)time, 0.00 %gc (2x)
I160315 20:50:39.621925 cli/start.go:183  initiating graceful shutdown of server
W160315 20:50:39.622644 storage/replica.go:1325  group 61 on store 1 failed to send message to 3: storage/raft_transport.go:255: node stopped
I160315 20:50:39.623086 storage/engine/rocksdb.go:152  closing rocksdb instance at "/mnt/data"
I160315 20:50:40.288935 cli/start.go:209  server drained and shutdown completed
```
[node0.log.parse.txt](https://github.com/cockroachdb/cockroach/files/174803/node0.log.parse.txt)

Node 1:
```
I160315 20:50:33.309370 storage/replica_command.go:1297  range 28: new leader lease replica {2 2 2} 2016-03-15 20:50:28.727894603 +0000 UTC 5.594562521s
I160315 20:50:40.286927 http2_client.go:877  transport: http2Client.notifyError got notified that the client transport was broken EOF.
I160315 20:50:40.287250 http2_client.go:877  transport: http2Client.notifyError got notified that the client transport was broken read tcp 172.31.58.173:45270->172.31.58.174:26257: read: connection reset by peer.
I160315 20:50:41.287719 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.174:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-174:26257"
I160315 20:50:41.288010 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.174:26257: getsockopt: connection refused"; Reconnecting to "ec2-54-209-133-121.compute-1.amazonaws.com:26257"
E160315 20:50:41.367549 server.go:2159  http: TLS handshake error from 65.209.47.35:50804: EOF
I160315 20:50:41.791429 cli/start.go:183  initiating graceful shutdown of server
I160315 20:50:41.792299 storage/engine/rocksdb.go:152  closing rocksdb instance at "/mnt/data"
I160315 20:50:42.166247 cli/start.go:209  server drained and shutdown completed
```
[node1.log.parse.txt](https://github.com/cockroachdb/cockroach/files/174806/node1.log.parse.txt)

Node 2:
```
I160315 20:50:35.822828 server/status/runtime.go:140  runtime stats: 93 goroutines, 7.71MiB active, 960.20cgo/sec, 0.02/0.00 %(u/s)time, 0.00 %gc (1x)
I160315 20:50:40.284518 http2_client.go:877  transport: http2Client.notifyError got notified that the client transport was broken EOF.
I160315 20:50:40.284544 http2_client.go:877  transport: http2Client.notifyError got notified that the client transport was broken EOF.
I160315 20:50:41.284828 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.174:26257: getsockopt: connection refused"; Reconnecting to "ec2-54-209-133-121.compute-1.amazonaws.com:26257"
I160315 20:50:41.285159 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.174:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-174:26257"
I160315 20:50:42.180275 http2_client.go:877  transport: http2Client.notifyError got notified that the client transport was broken EOF.
I160315 20:50:42.719875 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.174:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-174:26257"
I160315 20:50:42.720843 /go/src/google.golang.org/grpc/clientconn.go:501  grpc: Conn.transportMonitor exits due to: grpc: timed out trying to connect
I160315 20:50:42.726177 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.174:26257: getsockopt: connection refused"; Reconnecting to "ec2-54-209-133-121.compute-1.amazonaws.com:26257"
I160315 20:50:42.727014 /go/src/google.golang.org/grpc/clientconn.go:501  grpc: Conn.transportMonitor exits due to: grpc: timed out trying to connect
I160315 20:50:43.180559 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.173:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-173:26257"
I160315 20:50:43.683626 cli/start.go:183  initiating graceful shutdown of server
I160315 20:50:43.684410 storage/engine/rocksdb.go:152  closing rocksdb instance at "/mnt/data"
I160315 20:50:43.802606 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.174:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-174:26257"
I160315 20:50:43.802655 /go/src/google.golang.org/grpc/clientconn.go:318  Failed to dial ip-172-31-58-174:26257: grpc: the client connection is closing; please retry.
I160315 20:50:44.274443 cli/start.go:209  server drained and shutdown completed
```
[node2.log.parse.txt](https://github.com/cockroachdb/cockroach/files/174804/node2.log.parse.txt)

Node 3:
```
I160315 20:50:37.799474 server/status/runtime.go:140  runtime stats: 98 goroutines, 11.30MiB active, 1309.80cgo/sec, 0.02/0.00 %(u/s)time, 0.00 %gc (1x)
I160315 20:50:37.819908 storage/replica_command.go:1297  range 28: new leader lease replica {4 4 1} 2016-03-15 20:50:34.322457124 +0000 UTC 4.506650059s
I160315 20:50:40.286203 http2_client.go:877  transport: http2Client.notifyError got notified that the client transport was broken EOF.
I160315 20:50:40.286231 http2_client.go:877  transport: http2Client.notifyError got notified that the client transport was broken EOF.
I160315 20:50:41.286512 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.174:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-174:26257"
I160315 20:50:41.287192 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.174:26257: getsockopt: connection refused"; Reconnecting to "ec2-54-209-133-121.compute-1.amazonaws.com:26257"
I160315 20:50:42.181916 http2_client.go:877  transport: http2Client.notifyError got notified that the client transport was broken EOF.
I160315 20:50:42.709620 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.174:26257: getsockopt: connection refused"; Reconnecting to "ec2-54-209-133-121.compute-1.amazonaws.com:26257"
I160315 20:50:42.710826 /go/src/google.golang.org/grpc/clientconn.go:501  grpc: Conn.transportMonitor exits due to: grpc: timed out trying to connect
I160315 20:50:43.182234 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.173:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-173:26257"
I160315 20:50:43.201853 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.174:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-174:26257"
I160315 20:50:43.202812 /go/src/google.golang.org/grpc/clientconn.go:501  grpc: Conn.transportMonitor exits due to: grpc: timed out trying to connect
E160315 20:50:43.342584 server.go:2159  http: TLS handshake error from 65.209.47.35:50365: EOF
E160315 20:50:43.343754 server.go:2159  http: TLS handshake error from 65.209.47.35:50366: EOF
E160315 20:50:43.351303 server.go:2159  http: TLS handshake error from 65.209.47.35:50367: EOF
E160315 20:50:43.351603 server.go:2159  http: TLS handshake error from 65.209.47.35:6989: EOF
E160315 20:50:43.356242 server.go:2159  http: TLS handshake error from 65.209.47.35:50368: EOF
I160315 20:50:44.227705 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.174:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-174:26257"
I160315 20:50:44.292235 http2_client.go:877  transport: http2Client.notifyError got notified that the client transport was broken EOF.
I160315 20:50:44.973074 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.173:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-173:26257"
I160315 20:50:44.974265 /go/src/google.golang.org/grpc/clientconn.go:501  grpc: Conn.transportMonitor exits due to: grpc: timed out trying to connect
I160315 20:50:45.292559 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.172:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-172:26257"
I160315 20:50:45.777526 cli/start.go:183  initiating graceful shutdown of server
I160315 20:50:45.777699 stopper.go:352  draining; tasks left: 
5      server/node.go:701
1      storage/queue.go:380
E160315 20:50:45.778150 server/status.go:522  failed to send RPC: too many errors encountered (3 of 3 total): kv/send.go:264: rpc to ip-172-31-58-173:26257 failed as client connection was closed
E160315 20:50:45.778774 server/status.go:566  failed to send RPC: too many errors encountered (3 of 3 total): kv/send.go:264: rpc to ip-172-31-58-173:26257 failed as client connection was closed
W160315 20:50:45.827471 storage/replica.go:1325  group 1 on store 4 failed to send message to 2: storage/raft_transport.go:255: node stopped
W160315 20:50:45.827519 storage/replica.go:1325  group 1 on store 4 failed to send message to 1: storage/raft_transport.go:255: node stopped
I160315 20:50:45.961697 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.174:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-174:26257"
I160315 20:50:45.961738 /go/src/google.golang.org/grpc/clientconn.go:318  Failed to dial ip-172-31-58-174:26257: grpc: the client connection is closing; please retry.
I160315 20:50:46.028209 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.173:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-173:26257"
I160315 20:50:46.028242 /go/src/google.golang.org/grpc/clientconn.go:318  Failed to dial ip-172-31-58-173:26257: grpc: the client connection is closing; please retry.
W160315 20:50:46.627909 storage/replica.go:1325  group 7 on store 4 failed to send message to 3: storage/raft_transport.go:255: node stopped
I160315 20:50:46.746876 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.172:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-172:26257"
I160315 20:50:46.746909 /go/src/google.golang.org/grpc/clientconn.go:501  grpc: Conn.transportMonitor exits due to: grpc: the client connection is closing
I160315 20:50:50.777866 cli/start.go:195  6 running tasks
E160315 20:50:53.427099 server.go:503  http2: server: error reading preface from client 65.209.47.35:50392: timeout waiting for client preface
E160315 20:50:53.429100 server.go:503  http2: server: error reading preface from client 65.209.47.35:50393: timeout waiting for client preface
E160315 20:50:53.437599 server.go:503  http2: server: error reading preface from client 65.209.47.35:50398: timeout waiting for client preface
E160315 20:50:53.438271 server.go:503  http2: server: error reading preface from client 65.209.47.35:50399: timeout waiting for client preface
...
W160315 20:51:06.327389 storage/replica.go:1325  group 62 on store 4 failed to send message to 2: storage/raft_transport.go:262: queue for node 2 is full
W160315 20:51:06.327443 storage/replica.go:1325  group 4 on store 4 failed to send message to 2: storage/raft_transport.go:262: queue for node 16 is full
W160315 20:51:06.327496 storage/replica.go:1325  group 22 on store 4 failed to send message to 2: storage/raft_transport.go:262: queue for node 6 is full
W160315 20:51:06.427452 storage/replica.go:1325  group 42 on store 4 failed to send message to 2: storage/raft_transport.go:262: queue for node 5 is full
W160315 20:51:06.527767 storage/replica.go:1325  group 5 on store 4 failed to send message to 2: storage/raft_transport.go:262: queue for node 17 is full
W160315 20:51:06.527831 storage/replica.go:1325  group 1 on store 4 failed to send message to 2: storage/raft_transport.go:262: queue for node 8 is full
...
W160315 20:51:45.627498 storage/replica.go:1325  group 4 on store 4 failed to send message to 1: storage/raft_transport.go:262: queue for node 15 is full
W160315 20:51:45.627530 storage/replica.go:1325  group 4 on store 4 failed to send message to 2: storage/raft_transport.go:262: queue for node 16 is full
W160315 20:51:45.627574 storage/replica.go:1325  group 62 on store 4 failed to send message to 1: storage/raft_transport.go:262: queue for node 1 is full
W160315 20:51:45.627607 storage/replica.go:1325  group 62 on store 4 failed to send message to 2: storage/raft_transport.go:262: queue for node 2 is full
I160315 20:51:45.659131 /go/src/google.golang.org/grpc/clientconn.go:463  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 172.31.58.173:26257: getsockopt: connection refused"; Reconnecting to "ip-172-31-58-173:26257"
I160315 20:51:45.660210 /go/src/google.golang.org/grpc/clientconn.go:318  Failed to dial ip-172-31-58-173:26257: grpc: timed out trying to connect; please retry.
W160315 20:51:45.727392 storage/replica.go:1325  group 42 on store 4 failed to send message to 1: storage/raft_transport.go:262: queue for node 4 is full
W160315 20:51:45.727430 storage/replica.go:1325  group 42 on store 4 failed to send message to 2: storage/raft_transport.go:262: queue for node 5 is full
I160315 20:51:45.777722 cli/start.go:195  6 running tasks
E160315 20:51:45.777750 cli/start.go:206  time limit reached, initiating hard shutdown
```
[node3.log.parse.txt](https://github.com/cockroachdb/cockroach/files/174805/node3.log.parse.txt)

**Comment by tschottdorf** [comment]
The last node which shuts down doesn't have a working Raft group, so it will never shut down gracefully if any task which needs Raft (and doesn't have a timeout in its context) are in flight.

```
5      server/node.go:701
1      storage/queue.go:380
```

Must be one of these. Instead of investing time in figuring out a solution (unless it's simple, but I suspect this is going to crop up every now and then) maybe we should simply set the "hard shutdown" time limit lower and not show the message.

**Comment by jseldess** [comment]
Just another reason to fix this for beta: The **Getting Started** docs will provide copy/paste-ready commands for starting up three nodes locally and then shutting them down before securing the cluster. From what I understand, if this isn't fixed, we won't be able to tell users to shut down the third node with `cockroach quit`.

@spencerkimball

**Comment by tschottdorf** [comment]
Fixed in #7483.
